### ISSUE TITLE
Fix false INFINITELOOP on forever..mailbox.get()

### ISSUE THREAD

**Issue by svenka3** [verilator/verilator#4323]
As mailbox.get() is blocking task call, there is no INFINITELOOP in the code below.

```verilog
class bfm;
 mailbox #(int) in_box;
 int tr;
   
 function new();
  this.in_box = new (1);
 endfunction : new

  task main();
      // /* verilator lint_off INFINITELOOP */
    forever begin : loop
      in_box.get(tr);
      $display("%0t BFM received XACTN via MAILBOX as: %0d",
        $time, tr);
      //#1;
    end : loop
  endtask : main
endclass : bfm

module m;

  bfm bfm_0;

  initial begin
    #10;
    bfm_0 = new();
    fork
      bfm_0.main();
    join_none

    bfm_0.in_box.put(10);
    #10;
    bfm_0.in_box.put(20);
    #10;
    bfm_0.in_box.put(30);
    #10;
    $finish (2);
  end
endmodule : m

```
But verilator does throw:
```log
%Warning-INFINITELOOP: mbx_get_lint.sv:11:5: Infinite loop (condition always true)
                                           : ... In instance m
   11 |     forever begin : loop
      |     ^~~~~~~
                       ... For warning description see https://verilator.org/warn/INFINITELOOP?v=5.013
                       ... Use "/* verilator lint_off INFINITELOOP */" and lint_on around source to disable this message.
```
What 'verilator --version' are you using?  Did you try it with the git master version?

> Verilator 5.013 devel rev v5.008-304-gdfbd711e1 (mod)

What OS and distribution are you using?

> Windows WSL
May we assist you in trying to fix this in Verilator yourself?

> Yes, please

**Comment by wsnyder** [comment]
@svenka3 perhaps you could make a pull?  Basically thisWhileHasJumpDelay needs to be set on the mailbox operation.  At the point this is printed the mailbox operation has become an AstMethodCall, so the AstMethodCall class needs a new bool inside it to say that it has a delay inside it, then what creates that MethodCall needs to set that flag.

**Comment by svenka3** [comment]
Thanks, will attempt. Adding bool is done. I believe forever is translated to a while and hence I see this Warn coming from:
```cpp
    void visit(AstWhile* nodep) override {
```
In V3Const.cpp if I add:
```cpp
    void visit(AstMethodCall* nodep) override {
        if (nodep->name() == "get"/put etc.
        m_hasJumpDelay = true;
    }
```

Will that be a start? I would also need to enhance the condition:
```cpp
        m_hasJumpDelay = thisWhileHasJumpDelay || oldHasJumpDelay;
```

Regards

**Comment by svenka3** [comment]
On 
>inside it, then what creates that MethodCall needs to set that flag

```cpp
V3Width.cpp
 if (AstNodeFTask* const methodp = VN_CAST(foundp, NodeFTask)) {
                    nodep->replaceWith(new AstMethodCall{nodep->fileline(),
                                                         nodep->fromp()->unlinkFrBack(),
                                                         nodep->name(), nullptr});
```
That seems to be the place to set this flag? Am I looking at right place?

**Comment by wsnyder** [comment]
Yes, that is the right place.

**Comment by svenka3** [comment]
> Yes, that is the right place.

Thanks. Now within AstWhile method (where this check happens), how do I get handle to the container class to check for that flag? Also looking at V3Const.cpp

```cpp
const bool thisWhileHasJumpDelay = m_hasJumpDelay;
        m_hasJumpDelay = thisWhileHasJumpDelay || oldHasJumpDelay;
```
Further down it checks for:
```cpp
                if (!thisWhileHasJumpDelay) {

```
Curious - shouldn't we be checking if (!m_hasJumpDelay ) as that variable has the "OR-ed" result?

**Comment by kbieganski** [comment]
My 2 cents:

While checking for known blocking tasks fixes this specific case, it won't solve this bug in general.

Basically, the issue is when we're checking for infinite loops, we take a very superficial look at the statements in the loop. If there's a task call, we don't check if the task is blocking or not. So, even though `mailbox::get` does have a timing control inside (a `wait` statement), we cannot see it from the outside. This affects any and all blocking tasks.

One way to solve this is to peek inside, but then you have to iterate over the task's statements. More importantly, the property of 'being blocking' is contagious. If a task calls a blocking task, then it is also blocking. So we'd have to look further in and recursively check task calls.

Another way is to have some flag on the task that says if it's blocking. We have that for `AstCFunc`s at least, but that's set only after `V3Timing`. So even if we start flagging those earlier, there's the aforementioned problem of contagiousness.

Possibly the best solution is to delay detection of infinite loops until after `V3Timing`. At this point all blocking task calls are wrapped in `AstCAwait`s, so it's trivial to check if a call is blocking. Now there may be some things that break if we do that (or not, I have no idea), but I believe it's worth a try.

**Comment by wsnyder** [comment]
I tried delaying after V3Timing. the problem is some optimizations had already ripped out infinite loops, so they didn't get reported.

I tried doing twice:
1. Pre-timing, consider all task calls as potentially blocking so ignore any loops with them
2. Post-timing, more accurate as before.  But this still fired as could still be a call inside the while()

So, now just doing early as before, but suppressing whenever there's a task call, which might be too in-sensitive, we'll see.
